AsyncDisplayKit/ASMapNode.mm

```diff
@@ -0,0 +1,236 @@
+  _liveMap = liveMap;
+  liveMap ? [self addLiveMap] : [self removeLiveMap];
+}
+
+- (CGSize)mapSize
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _mapSize;
+}
+
+- (void)setMapSize:(CGSize)mapSize
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (CGSizeEqualToSize(mapSize,_mapSize)) {
+    return;
+  }
+  _mapSize = mapSize;
+  _nodeSize = _mapSize;
+  _automaticallyReloadsMapImageOnOrientationChange = NO;
+  [self setNeedsLayout];
+}
+
+- (BOOL)automaticallyReloadsMapImageOnOrientationChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  return _automaticallyReloadsMapImageOnOrientationChange;
+}
+
+- (void)setAutomaticallyReloadsMapImageOnOrientationChange:(BOOL)automaticallyReloadsMapImageOnOrientationChange
+{
+  ASDN::MutexLocker l(_propertyLock);
+  if (_automaticallyReloadsMapImageOnOrientationChange == automaticallyReloadsMapImageOnOrientationChange) {
+    return;
+  }
+  _automaticallyReloadsMapImageOnOrientationChange = automaticallyReloadsMapImageOnOrientationChange;
+  
+}
+
+- (void)fetchData
+{
+  [super fetchData];
+  [self setUpSnapshotter];
+  [self takeSnapshot];
+}
+
+- (void)clearFetchedData
+{
+  [super clearFetchedData];
+  if (_mapView) {
+    [_mapView removeFromSupernode];
+    _mapView = nil;
+  }
+  _mapImage.image = nil;
+}
+
+- (void)takeSnapshot
+{
+  if (!_snapshotter.isLoading) {
+    [_snapshotter startWithCompletionHandler:^(MKMapSnapshot *snapshot, NSError *error) {
+      if (!error) {
+        UIImage *image = snapshot.image;
+        CGRect finalImageRect = CGRectMake(0, 0, image.size.width, image.size.height);
+        
+        // Get a standard annotation view pin. Future implementations should use a custom annotation image property.
+        MKAnnotationView *pin = [[MKPinAnnotationView alloc] initWithAnnotation:nil reuseIdentifier:@""];
+        UIImage *pinImage = pin.image;
+        
+        UIGraphicsBeginImageContextWithOptions(image.size, YES, image.scale);
+        [image drawAtPoint:CGPointMake(0, 0)];
+        
+        for (id<MKAnnotation>annotation in _annotations)
+        {
+          CGPoint point = [snapshot pointForCoordinate:annotation.coordinate];
+          if (CGRectContainsPoint(finalImageRect, point))
+          {
+            CGPoint pinCenterOffset = pin.centerOffset;
+            point.x -= pin.bounds.size.width / 2.0;
+            point.y -= pin.bounds.size.height / 2.0;
+            point.x += pinCenterOffset.x;
+            point.y += pinCenterOffset.y;
+            [pinImage drawAtPoint:point];
+          }
+        }
+        UIImage *finalImage = UIGraphicsGetImageFromCurrentImageContext();
+        UIGraphicsEndImageContext();
+        _mapImage.image = finalImage;
+      }
+    }];
+  }
+}
+
+- (void)resetSnapshotter
+{
+  if (!_snapshotter.isLoading) {
+    _options.size = _nodeSize;
+    _snapshotter = [[MKMapSnapshotter alloc] initWithOptions:_options];
+  }
+}
+
+#pragma mark - Action
+- (void)addLiveMap
+{
+  if (self.isNodeLoaded && !_mapView) {
```

Comment:
-addLiveMap is only called from -setLiveMap: when _isLiveMap = YES. So it is set beforehand and assumes that _isLiveMap is set to YES. Since this is an internal method, I would say that would suffice. Or do you prefer to be more defensive in this regard?